auth/client.go

```diff
@@ -67,7 +69,32 @@ func (t *HTTPTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	base := t.opts.Base
 	t.mu.Unlock()
 
-	resp, err := base.RoundTrip(req)
+	// req1 is our first request in the authorization flow.
+	//
+	// If we mutate its body, we must clone it first.
+	req1 := req
+	var (
+		// If haveBody is set, the request has a nontrivial body, and need avoid
+		// reading (or closing) it multiple times. In that case, bodyBytes is its
+		// content.
+		haveBody  bool
```

Comment:
Couldn't you just test whether bodyBytes is non-nil?

Reply:
I'd rather just have a variable keep track of whether we mutated, than rely on some convention of nillness vs emptyness.

Strictly speaking, the body could be empty but have a non-trivial Close function, and we should be very precise here.